internal/datastore/mysql/datastore.go

```diff
@@ -0,0 +1,648 @@
+		gcCtx:                    gcCtx,
+		cancelGc:                 cancelGc,
+		watchBufferLength:        config.watchBufferLength,
+		createTxn:                createTxn,
+		createBaseTxn:            createBaseTxn,
+		QueryBuilder:             queryBuilder,
+		querySplitter:            &querySplitter,
+	}
+
+	// Start a goroutine for garbage collection.
+	if store.gcInterval > 0*time.Minute {
+		store.gcGroup, store.gcCtx = errgroup.WithContext(store.gcCtx)
+		store.gcGroup.Go(store.runGarbageCollector)
+	} else {
+		log.Warn().Msg("garbage collection disabled in mysql driver")
+	}
+
+	return store, nil
+}
+
+// unlike PSQL, this implementation does not create a transaction - not needed
+// prepared statements are also not used given they perform poorly
+// on environments where connections have short lifetime (e.g.
+// to gracefully handle load-balancer connection drain)
+func newMySQLExecutor(db *sql.DB) common.ExecuteQueryFunc {
+	return func(ctx context.Context, revision datastore.Revision, sqlQuery string, args []interface{}) ([]*core.RelationTuple, error) {
+		ctx = datastore.SeparateContextWithTracing(ctx)
+
+		span := trace.SpanFromContext(ctx)
+
+		rows, err := db.QueryContext(ctx, sqlQuery, args...)
+		if err != nil {
+			return nil, fmt.Errorf(errUnableToQueryTuples, err)
+		}
+		defer migrations.LogOnError(ctx, rows.Close)
+
+		span.AddEvent("Query issued to database")
+
+		var tuples []*core.RelationTuple
+		for rows.Next() {
+			nextTuple := &core.RelationTuple{
+				ObjectAndRelation: &core.ObjectAndRelation{},
+				User: &core.User{
+					UserOneof: &core.User_Userset{
+						Userset: &core.ObjectAndRelation{},
+					},
+				},
+			}
+			userset := nextTuple.User.GetUserset()
+			err := rows.Scan(
+				&nextTuple.ObjectAndRelation.Namespace,
+				&nextTuple.ObjectAndRelation.ObjectId,
+				&nextTuple.ObjectAndRelation.Relation,
+				&userset.Namespace,
+				&userset.ObjectId,
+				&userset.Relation,
+			)
+			if err != nil {
+				return nil, fmt.Errorf(errUnableToQueryTuples, err)
+			}
+
+			tuples = append(tuples, nextTuple)
+		}
+		if err := rows.Err(); err != nil {
+			return nil, fmt.Errorf(errUnableToQueryTuples, err)
+		}
+		span.AddEvent("Tuples loaded", trace.WithAttributes(attribute.Int("tupleCount", len(tuples))))
+		return tuples, nil
+	}
+}
+
+type Datastore struct {
+	db            *sql.DB
+	driver        *migrations.MysqlDriver
+	querySplitter *common.TupleQuerySplitter
+	url           string
+
+	revisionFuzzingTimedelta time.Duration
+	gcWindowInverted         time.Duration
+	gcInterval               time.Duration
+	gcMaxOperationTime       time.Duration
+	watchBufferLength        uint16
+
+	gcGroup  *errgroup.Group
+	gcCtx    context.Context
+	cancelGc context.CancelFunc
+
+	createTxn     string
+	createBaseTxn string
+
+	*QueryBuilder
+}
+
+// Close closes the data store.
+func (mds *Datastore) Close() error {
+	// FIXME dupe from postgres datastore - need to refactor
```

Comment:
// TODO(yourgithubusername)

Reply:
addressed in https://github.com/authzed/spicedb/pull/525/commits/4dcb58c11caa4b9fd9fffe6ffe8b153108dc7d10